geo_tag/geo_tag.py

```diff
@@ -0,0 +1,200 @@
+            geo_locator = Nominatim()
+            location = geo_locator.geocode(target_key)
+            self.coord = (location.longitude, location.latitude)
+
+        geo_content = self.cache.get(target_key)
+        if geo_content is not None:
+            geo_tuple = dict()
+            geo_tuple["stateID"] = geo_content[2]
+            geo_tuple["stateName"] = full_state_name
+            geo_tuple["countyID"] = geo_content[1]
+            geo_tuple["countyName"] = geo_content[3]
+            geo_tuple["cityID"] = geo_content[0]
+            geo_tuple["cityName"] = city_state_name.split(",")[0]
+            geo_tuple["coordinate"] = self.coord
+            geo_tuple["source"] = flag
+
+            self.data['geo_tag'] = geo_tuple
+            print(geo_tuple)
+            
+        else:
+            print("[ERROR] Not find the county of", target_key)
+
+    def infer_geo_from_place(self):
+        place = self.data.get('place')
+        if place is not None:
+            # extract the city and state abbrev
+            city_state_name = place.get('full_name')
+            self.extract_geo_tag_from_city_and_state(city_state_name, "place")
+
+    def infer_geo_from_coord(self):
+        geo_locator = Nominatim()
+        # geopy require (Lat, Long)
+        t = (self.coord[1], self.coord[0])
+        location = geo_locator.reverse(t)
+
+        if location is None:
+            self.data['geo_tag'] = {"coordinate": self.coord}
+            return
+
+        # make up the (city, state) as key to search
+        target_key = location.address.split(",")[3].strip() + ', ' + location.address.split(",")[5].strip()
+        geo_content = self.cache.get(target_key)
+
+        if geo_content is not None:
+            geo_tuple = dict()
+            geo_tuple["stateID"] = geo_content[2]
+            geo_tuple["stateName"] = location.address.split(",")[5].strip()
+            geo_tuple["countyID"] = geo_content[1]
+            geo_tuple["countyName"] = geo_content[3]
+            geo_tuple["cityID"] = geo_content[0]
+            geo_tuple["cityName"] = location.address.split(",")[3].strip()
+            geo_tuple["coordinate"] = self.coord
+            geo_tuple["source"] = "coordinate"
+
+            self.data['geo_tag'] = geo_tuple
+            print(geo_tuple)
+
+        else:
+            print("[ERROR] Not find the county of", target_key)
+            self.data['geo_tag'] = {"coordinate": self.coord}
+
+    def infer_geo_from_user(self):
+        user = self.data.get('user')
+        if user is not None:
+            # extract the city and state abbrev
+            city_state_name = user.get('location')
+            self.extract_geo_tag_from_city_and_state(city_state_name, "user")
+        else:
+            # The methods above are all not working, for now we skip.
+            # TODO: Use NLP to infer geo_tag from text in tweet.
+            self.data['geo_tag'] = dict()
+
+    def tag_one_tweet(self, tweet_json, city_json):
```

Comment:
I would suggest to make this method a static method. Avoid using `self`, make it fully functional.

Reply:
But we need many `self` related variables and private functions which make me think is may not be a static method?